docs/architecture/010-transactions.md

```diff
@@ -0,0 +1,63 @@
+# 010: Database transactions
+
+Authors: @opqdonut
+
+## Introduction
+
+This is an ADR written in aftermath to a production incident to
+describe our current design choices with database transactions.
+
+## Design philosophy
+
+At this point in the software lifecycle of REMS, we err on the side of
+caution and try to guarantee consistency of data at the cost of
+performance.
+
+## Every API call is a transaction
+
+The `rems.api/transaction-middleware` middleware wraps each API call
+in a transaction. The transactions are run with isolation level
+serializable, the strictest serialization level (see [PostgreSQL docs]
+for more). Additionally, the transactions for GET requests are
+read-only to guarantee that our GETs are pure.
+
+[PostgreSQL docs]: https://www.postgresql.org/docs/9.6/transaction-iso.html
+
+## Strict ordering of commands via locking
+
+Our application commands are processed roughly in the following way
+(implementation: `rems.api.services.command/command!`):
+
+1. get all events for application from the `application_event` table
+2. compute application state
+3. run command handler (given command and application state)
+4. if command handler succeeds, append new events to `application_event`
+5. run process managers on new events, which can result in
+   - new commands (e.g. from `approver-bot`)
+   - new rows in db (e.g. `outbox` or `blacklist` rows)
```

Comment:
>>>
These commands are processed immediately, in the same thread.

Reply:
done